barretenberg/cpp/src/barretenberg/vm2/tracegen/lib/phase_spec.cpp

```diff
@@ -0,0 +1,83 @@
+          /*length_offset=*/AVM_PUBLIC_INPUTS_PUBLIC_CALL_REQUEST_ARRAY_LENGTHS_APP_LOGIC_CALLS_ROW_IDX,
+      } },
+    { TransactionPhase::TEARDOWN,
+      {
+          /*read_offset=*/AVM_PUBLIC_INPUTS_PUBLIC_TEARDOWN_CALL_REQUEST_ROW_IDX,
+          /*write_offset=*/0, // Write offset is 0 for public call requests
+          /*length_offset=*/AVM_PUBLIC_INPUTS_PUBLIC_CALL_REQUEST_ARRAY_LENGTHS_TEARDOWN_CALL_ROW_IDX,
+      } },
+    { TransactionPhase::NR_NOTE_INSERTION,
+      {
+          /*read_offset=*/AVM_PUBLIC_INPUTS_PREVIOUS_NON_REVERTIBLE_ACCUMULATED_DATA_NOTE_HASHES_ROW_IDX,
+          /*write_offset=*/AVM_PUBLIC_INPUTS_AVM_ACCUMULATED_DATA_NOTE_HASHES_ROW_IDX,
+          /*length_offset=*/AVM_PUBLIC_INPUTS_PREVIOUS_NON_REVERTIBLE_ACCUMULATED_DATA_ARRAY_LENGTHS_ROW_IDX,
+      } },
+    { TransactionPhase::NR_NULLIFIER_INSERTION,
+      {
+          /*read_offset=*/AVM_PUBLIC_INPUTS_PREVIOUS_NON_REVERTIBLE_ACCUMULATED_DATA_NULLIFIERS_ROW_IDX,
+          /*write_offset=*/AVM_PUBLIC_INPUTS_AVM_ACCUMULATED_DATA_NULLIFIERS_ROW_IDX,
+          /*length_offset=*/AVM_PUBLIC_INPUTS_PREVIOUS_NON_REVERTIBLE_ACCUMULATED_DATA_ARRAY_LENGTHS_ROW_IDX,
+      } },
+    { TransactionPhase::R_NOTE_INSERTION,
+      {
+          /*read_offset=*/AVM_PUBLIC_INPUTS_PREVIOUS_REVERTIBLE_ACCUMULATED_DATA_NOTE_HASHES_ROW_IDX,
+          /*write_offset=*/AVM_PUBLIC_INPUTS_AVM_ACCUMULATED_DATA_NOTE_HASHES_ROW_IDX,
+          /*length_offset=*/AVM_PUBLIC_INPUTS_PREVIOUS_REVERTIBLE_ACCUMULATED_DATA_ARRAY_LENGTHS_ROW_IDX,
+      } },
+    { TransactionPhase::R_NULLIFIER_INSERTION,
+      {
+          /*read_offset=*/AVM_PUBLIC_INPUTS_PREVIOUS_REVERTIBLE_ACCUMULATED_DATA_NULLIFIERS_ROW_IDX,
+          /*write_offset=*/AVM_PUBLIC_INPUTS_AVM_ACCUMULATED_DATA_NULLIFIERS_ROW_IDX,
+          /*length_offset=*/AVM_PUBLIC_INPUTS_PREVIOUS_REVERTIBLE_ACCUMULATED_DATA_ARRAY_LENGTHS_ROW_IDX,
+      } },
+    { TransactionPhase::NR_L2_TO_L1_MESSAGE,
+      {
+          /*read_offset=*/AVM_PUBLIC_INPUTS_PREVIOUS_NON_REVERTIBLE_ACCUMULATED_DATA_L2_TO_L1_MSGS_ROW_IDX,
+          /*write_offset=*/AVM_PUBLIC_INPUTS_AVM_ACCUMULATED_DATA_L2_TO_L1_MSGS_ROW_IDX,
+          /*length_offset=*/AVM_PUBLIC_INPUTS_PREVIOUS_NON_REVERTIBLE_ACCUMULATED_DATA_ARRAY_LENGTHS_ROW_IDX,
+      } },
+    { TransactionPhase::R_L2_TO_L1_MESSAGE,
+      {
+          /*read_offset=*/AVM_PUBLIC_INPUTS_PREVIOUS_REVERTIBLE_ACCUMULATED_DATA_L2_TO_L1_MSGS_ROW_IDX,
+          /*write_offset=*/AVM_PUBLIC_INPUTS_AVM_ACCUMULATED_DATA_L2_TO_L1_MSGS_ROW_IDX,
+          /*length_offset=*/AVM_PUBLIC_INPUTS_PREVIOUS_REVERTIBLE_ACCUMULATED_DATA_ARRAY_LENGTHS_ROW_IDX,
+      } },
+
+} };
+} // namespace
+
+const TxPhaseOffsetsTable::Offsets& TxPhaseOffsetsTable::get_offsets(TransactionPhase phase) const
+{
+    return PHASE_PUB_INPUTS_OFFSETS.at(phase);
+}
+
+const TxPhaseOffsetsTable& TxPhaseOffsetsTable::get()
+{
+    static const TxPhaseOffsetsTable table;
```

Comment:
You don't need a singleton. You can put the map above inside a function that returns it.

Then you do
```
const std::unordered_map<TransactionPhase, TxPhaseOffsetsTable::Offsets> PHASE_PUB_INPUTS_OFFSETS = generate_offsets();
```

Reply:
Actually this is also not needed, how you create the PHASE_PUB_INPUTS_OFFSETS is ok. You still don't strictly need a singleton for this :)